automation: Add support for tests

ricekot opened the pull request:
- Only stats tests are supported for now.
- Part of zaproxy/zaproxy#6496

<details>
<summary> Simple UML diagram </summary>
<img src="https://user-images.githubusercontent.com/16446369/116572992-0efb4180-a92a-11eb-85ba-b998f817ec3b.png">
</details>

Signed-off-by: ricekot <ricekot@gmail.com>

thc202 commented:
Guess this should wait for #2888 (or the other way around), I'd expect this to also have verification.

It could have some tests.

psiinon commented:
I think we should add tests to the min and max generated spider jobs etc for the number of urls found - these would have an action of "ignore" as we wouldnt know what to expect, of course. These tests would replace the existing `failIfFoundUrlsLessThan` etc params. Not sure if those params should be removed or marked as deprecated in some way?
We could 'warn' on their use stating that the new tests should be used instead.

ricekot commented:
Updated.

psiinon commented on addOns/automation/src/main/javahelp/org/zaproxy/addon/automation/resources/help/contents/job-spider.html:
```diff
@@ -39,6 +37,13 @@ <H2>YAML</H2>
       sendRefererHeader:               # Bool: Whether the spider will send the referer header, default: true
       threadCount:                     # Int: The number of spider threads, default: 2
       userAgent:                       # String: The user agent to use in requests, default: '' - use the default ZAP one 
+    tests:
```
I dont think we can expect the users to know the ZAP stats or which ones would be most relevant. So for both the help and the min/max data I think we should include the tests we think are most appropriate (if any) - so in this case it could be something like:
```
- name: 'Check number of URLs found'   # String: Name of the test, default: statistic + operator + value
  type: 'stats'                        # String: Type of test, supported: 'stats'
  statistic: 'spider.urls.added'       # String: Name of an integer / long statistic, supported: 'spider.urls.added'
  operator: '<'                        # String ['==', '!=', '&gt;=', '&gt;', '&lt;', '&lt;=']: Operator used for testing
  value: 100                           # Int: Value to compare statistic against, change this to the min number of urls you expect to find
  onFail: 'ignore'                     # String [warn, error, ignore]: Change this to 'warn' or 'error' for the test to take effect
```

psiinon commented on addOns/automation/src/main/javahelp/org/zaproxy/addon/automation/resources/help/contents/tests.html:
```diff
@@ -0,0 +1,35 @@
+
+<H2>Statistics</H2>
+Statistics-based tests are universally supported by all add-ons that have an automation job. If there is a relevant
+statistic used by an add-on, a test can be created with it.
```
Note for the team - should we document all the stats somewhere eg on the website??

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AutomationStatisticTest.java:
```diff
@@ -0,0 +1,82 @@
+                progress.error(testFailedMessage);
+                break;
+            case "ignore":
+                break;
```
progress.info ?

psiinon commented on addOns/spiderAjax/CHANGELOG.md:
```diff
@@ -7,6 +7,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Now using 2.10 logging infrastructure (Log4j 2.x).
 - Maintenance changes.
+- Remove automation parameters `failIfFoundUrlsLessThan` and `warnIfFoundUrlsLessThan`
+  in favour of the `spiderAjax.urls.added` statistic test.
```
The add-on will also need to depend on the new version on the automation add-on.

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AutomationStatsListener.java:
```diff
@@ -0,0 +1,108 @@
+/*
```
I'm not sure this class is needed. We can reply in the ZAP in-memory stats as they record exactly the same things. If they have been disabled then we can just warn that the tests wont work.

psiinon commented on addOns/spiderAjax/src/main/javahelp/org/zaproxy/zap/extension/spiderAjax/resources/help/contents/automation.html:
```diff
@@ -29,6 +27,14 @@ <H2>Job: spiderAjax</H2>
       maxCrawlStates:                  # Int: The maximum number of crawl states the crawler should crawl, default: 0 unlimited
       randomInputs:                    # Bool: When enabled random values will be entered into input element, default: true
       reloadWait:                      # Int: The time in millseconds to wait after the URL is loaded, default: 1000
+    tests:
```
I think it would be worth adding a generic tests help page which would explain that you can add as many tests as you like to any jobs. We could then link from there to more info about the stats, if we create that sometime :P

psiinon reviewed (changes requested):
Getting closer :)

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AutomationStatisticTest.java:
```diff
@@ -0,0 +1,97 @@
+
+    public void logToProgressIfFailed(AutomationProgress progress) throws RuntimeException {
+        if (!hasFailed()) {
+            return;
```
Log to info that its passed? Lets the user know the test was valid and may give additional useful feedback if we state the exact result, eg something like `Job myjob Test passed stats.aaa.bbb 123 >= 100`

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/SpiderJob.java:
```diff
@@ -42,14 +43,11 @@
 
-    private int failIfFoundUrlsLessThan = 0;
-    private int warnIfFoundUrlsLessThan = 0;
+    private AutomationStatisticTest urlsFoundTest;
```
I'm confused - why do we need this? I was expecting this to just be a generic test so no specific code in the job :/

ricekot commented:
Which of the two do we prefer?

```yaml
name: 'Fail if URLs < 100'
type: 'stats'
statistic: 'spider.urls.added'
operator: '>='
value: 100
onFail: 'error'
```

```yaml
name: 'Fail if URLs < 100'
type: 'stats'
statistic: 'spider.urls.added'
operator: '<'
value: 100
action: 'error'
```

The only difference between the two is the operator. My question is, should we fail if the given condition is satisfied, or if it is not?

thc202 commented:
If it's not (IMO).

psiinon commented:
Agreed.
Tests should be positive- Test that the spider finds more than X urls. Then fail if the test is negative.

ricekot commented:
Updated and ready for a review!

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/SpiderJob.java:
```diff
@@ -228,22 +208,11 @@ public void runJob(
+                .getExtensionLoader()
+                .getExtension(ExtensionStats.class)
+                .getInMemoryStats()
+                .counterInc("spider.urls.added", numUrlsFound);
```
Ideally we'll add full stats to the core spider. In which case it might be better to associate this stats name with the automation job so that the distinction is clear, eg something like `automation.spider.urls.added`?

psiinon commented on addOns/spiderAjax/src/main/java/org/zaproxy/zap/extension/spiderAjax/automation/AjaxSpiderJob.java:
```diff
@@ -184,22 +172,11 @@ public void runJob(
+                .getExtensionLoader()
+                .getExtension(ExtensionStats.class)
+                .getInMemoryStats()
+                .counterInc("spiderAjax.urls.added", numUrlsFound);
```
Ideally this would be done in the main ajax spider code so that its incremented as urls are added. Could be a good time to add other stats (like out of scope requests) but could be left for another PR (which someone else could do).

psiinon reviewed (changes requested):
Relatively minor changes.

psiinon commented:
Also, we should probably have an abstract AutomationTest that the AutomationJob uses and thats implemented by AutomationStatisticsTest. This will make it easier for us to add new types of tests later.

thc202 commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AbstractAutomationTest.java:
```diff
@@ -0,0 +1,69 @@
+        this.onFail = onFail;
+    }
+
+    public void logToProgressIfFailed(AutomationProgress progress) throws RuntimeException {
```
Just `logToProgress`? It also logs when passed.

thc202 commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AbstractAutomationTest.java:
```diff
@@ -0,0 +1,69 @@
+            return;
+        }
+        switch (onFail) {
+            case "warn":
```
Could use constants (here and for the collection, same for the operators).

thc202 replied:
Sorry, just noticed your message. I'm fine with the enum as well.

thc202 commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AutomationJob.java:
```diff
@@ -282,6 +284,66 @@ void verifyOrApplyParameters(
+                .forEach(t -> t.logToProgressIfFailed(progress));
+    }
+
+    private <T> T safeCast(Object property, Class<T> clazz) {
```
`static`

thc202 commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/SpiderJob.java:
```diff
@@ -228,22 +208,11 @@ public void runJob(
+        Control.getSingleton()
+                .getExtensionLoader()
+                .getExtension(ExtensionStats.class)
+                .getInMemoryStats()
```
These (Ajax Spider) could also lead to NPE.

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/jobs/SpiderJob.java:
```diff
@@ -51,15 +52,10 @@
     private static final String PARAM_CONTEXT = "context";
     private static final String PARAM_URL = "url";
-    private static final String PARAM_FAIL_IF_LESS_URLS = "failIfFoundUrlsLessThan";
-    private static final String PARAM_WARN_IF_LESS_URLS = "warnIfFoundUrlsLessThan";
```
Should we keep these (for now), but warn if they are used and not actually use them?
In theory people could already be using them which would mean their jobs would start to fail. Warning on then would mean the jobs carry on working unless they are set to fail-on-warn.

ricekot replied:
Okay, we decided on IRC earlier to remove them, but I'll add them back again.

psiinon replied:
I'm also ok with that, is was really a question :)

pranavsaxena17 commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/ExtensionAutomation.java:
```diff
@@ -240,6 +252,7 @@ private AutomationProgress runPlan(LinkedHashMap<?, ?> envData, ArrayList<?> job
                 break;
             }
             progress.info(Constant.messages.getString("automation.info.jobend", job.getType()));
+            job.logFailedTestsToProgress(progress);
```
I think this should be before the `env.isTimeToQuit()` check, otherwise the next job would be run before checking whether the previous one had any failed tests logged to the progress which should have lead to quitting the plan.

thc202 commented:
Thank you!

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AutomationJob.java:
```diff
@@ -36,6 +37,7 @@
 
     private String name;
     private AutomationEnvironment env;
+    private final Map<String, List<AutomationStatisticTest>> statsMap = new HashMap<>();
```
Shouldnt this be AbstractAutomationTest ?

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AutomationJob.java:
```diff
@@ -282,6 +284,73 @@ void verifyOrApplyParameters(
+            }
+            LinkedHashMap<?, ?> testData = (LinkedHashMap<?, ?>) testObj;
+            Object testType = testData.get("type");
+            if ("stats".equals(testType)) {
```
Warn if unknown testType? Will also catch typos :)

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AutomationJob.java:
```diff
@@ -282,6 +284,73 @@ void verifyOrApplyParameters(
+        }
+    }
+
+    public void addStatisticTest(AutomationStatisticTest test) {
```
Should be `addTest(AbstractAutomationTest test)` ?

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AutomationJob.java:
```diff
@@ -282,6 +284,73 @@ void verifyOrApplyParameters(
+        return null;
+    }
+
+    Map<String, List<AutomationStatisticTest>> getStatsMap() {
```
`List<AbstractAutomationTest>`

psiinon commented on addOns/automation/src/main/java/org/zaproxy/addon/automation/AutomationStatisticTest.java:
```diff
@@ -0,0 +1,145 @@
+                        .getExtension(ExtensionStats.class)
+                        .getInMemoryStats();
+        stat =
+                inMemoryStats != null
```
I'd say we should always warn if there are no in-memory stats

psiinon replied:
We could always do this when adding the stats test?

psiinon reviewed (changes requested):
All pretty minor :)

kingthorin commented:
Whoop whooop! Nice work @ricekot ❗
